ddtrace/tracer/payload.go

```diff
@@ -108,9 +113,20 @@ func (p *payload) updateHeader() {
 func (p *payload) Read(b []byte) (n int, err error) {
 	if p.off < len(p.header) {
 		// reading header
-		n = copy(b, p.header[p.off:])
+		n, err := p.buf.Write(p.header[p.off:])
+		if err != nil {
+			return n, err
+		}
 		p.off += n
-		return n, nil
+	}
+	for len(p.traces) != 0 && p.buf.Len() <= len(b) {
```

Comment:
I think we'll still have a problem here. If the condition is met and `p.buf.Len() <= len(b)` and a trace is written to `p.buf` but `len(b)` can not accommodate it in its entirety, it may be read only partially during a crash or restart, which means that on subsequent retries again "Bad Request" will be returned because the msgpack payload will not be decodable.

Reply:
I'd still be happy considering this implementation however and just dropping any retry policy, considering any traces sent while the agent is down lost.

Reply:
That's true, I didn't think about this case. I think we could still keep the retry policy but extend on it, by adding a method to the payload for resetting the offset and buffer, which would make the payload readable again on subsequent flush retries, while not losing any traces still present in the payload.